gpt4all-chat/database.cpp

```diff
@@ -948,40 +1163,100 @@ void Database::start()
     if (m_embeddings->fileExists() && !m_embeddings->load())
         qWarning() << "ERROR: Could not load embeddings";
 
-    int nAdded = addCurrentFolders();
-    Network::globalInstance()->trackEvent("localdocs_startup", { {"doc_collections_total", nAdded} });
+    addCurrentFolders();
 }
 
-int Database::addCurrentFolders()
+void Database::addCurrentFolders()
 {
 #if defined(DEBUG)
     qDebug() << "addCurrentFolders";
 #endif
 
     QSqlQuery q;
     QList<CollectionItem> collections;
-    if (!selectAllFromCollections(q, &collections)) {
+    if (!selectAllFromCollections(LOCALDOCS_VERSION, q, &collections)) {
         qWarning() << "ERROR: Cannot select collections" << q.lastError();
-        return 0;
+        return;
     }
 
-    emit collectionListUpdated(collections);
+    guiCollectionListUpdated(collections);
 
-    int nAdded = 0;
-    for (const auto &i : collections)
-        nAdded += addFolder(i.collection, i.folder_path, true);
+    for (const auto &i : collections) {
+        if (!i.forceIndexing) {
+            scheduleUncompletedEmbeddings(i.folder_id);
+            addFolder(i.collection, i.folder_path);
+        }
+    }
 
-    updateIndexingStatus();
+    updateCollectionStatistics();
+}
+
+void Database::scheduleUncompletedEmbeddings(int folder_id)
+{
+    QList<EmbeddingChunk> chunkList;
+    QSqlQuery q;
+    if (!selectAllUncompletedChunks(q, folder_id, &chunkList)) {
+        qWarning() << "ERROR: Cannot select uncompleted chunks" << q.lastError();
+        return;
+    }
+
+    if (chunkList.isEmpty())
+        return;
 
-    return collections.count() + nAdded;
+    int total = 0;
+    if (!selectCountChunks(q, folder_id, &total)) {
+        qWarning() << "ERROR: Cannot count total chunks" << q.lastError();
+        return;
+    }
+
+    CollectionItem item = guiCollectionItem(folder_id);
+    item.totalEmbeddingsToIndex = total;
+    item.currentEmbeddingsToIndex = total - chunkList.size();
+    updateGuiForCollectionItem(item);
+
+    const int batchSize = 100;
+    for (int i = 0; i < chunkList.size(); i += batchSize) {
+        QList<EmbeddingChunk> batch = chunkList.mid(i, qMin(batchSize, chunkList.size() - i));
```

Comment:
According to the [docs](https://doc.qt.io/qt-6/qlist.html#mid), this can just be:
```suggestion
        QList<EmbeddingChunk> batch = chunkList.mid(i, batchSize);
```
> otherwise length elements **(or all remaining elements if there are less than length elements)** are included.

Reply:
fixed.